lib/Sema/ConstraintGraph.cpp

```diff
@@ -980,6 +980,17 @@ namespace {
             visitedConstraints);
       }
 
+      for (auto &constraint : cs.getConstraints()) {
+        if (constraint.getKind() == ConstraintKind::Disjunction ||
+            constraint.getKind() == ConstraintKind::Conjunction) {
```

Comment:
Depth-first search should visit all of constraints and type variables, shouldn't it? I actually don't know why we had go through (all) constraints before to find type variables, that stands out to me, all of the "in scope" type variables should already be in "typeVars" list passed to the `ConnectedComponents`. It seems like `validComponents` is `representatives` really...

Reply:
We omit type variables from validComponents if they’re fixed to a concrete type. I reverse engineered the old behavior which would unconditionally walk all constraints as well, it’s actually sufficient to just consider disjunctions and conjunctions.

I wasn’t able to figure out why disjunctions and conjunctions should introduce a new component even if the type variable has been fixed. It might be a bug elsewhere.

if we can get rid of this loop over constraints it speeds things up further!

Reply:
Would you mind figuring out why conjunction/disjunctions are special? My fear is that they are not the only ones but we just don’t have test cases to cover over cases like constraints with additional type variable references. Logically I think we don’t need valid components at all and representatives should be sufficient and most likely depth-first search needs to walk into type variables that are bound or something like that to cover all corner cases.

Reply:
Sure, I'll take a look

Reply:
Thank you!